### PR TITLE
Fix more inefficient PT operations

### PR DESCRIPTION
**Pull request by cyyever**
Remove ``clone().detach()`` and unnecessary ``cpu()`` calls.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Hi @cyyever, is `detach()` even necessary before `tolist()`? Just calling `tolist()` on a tensor, even one with grad enabled and on the CUDA device, seems to work for me.

**Comment by cyyever** [comment]
@Rocketknight1 I forgot. If you can work on your host, they are not needed.

File: src/transformers/models/rag/modeling_rag.py [hunk 1]
```diff
@@ -981,7 +981,7 @@ def generate(
             question_hidden_states = self.question_encoder(input_ids, attention_mask=attention_mask)[0]
             context_input_ids = self.retriever(
                 input_ids,
-                question_hidden_states.cpu().detach().to(torch.float32).numpy(),
+                question_hidden_states.detach().to(devoce="cpu", dtype=torch.float32).numpy(),
```
Comment (gante) [inline src/transformers/models/rag/modeling_rag.py:984]:
```suggestion
                question_hidden_states.detach().to(device="cpu", dtype=torch.float32).numpy(),
```
:D

File: src/transformers/agents/agent_types.py [hunk 2]
```diff
@@ -129,7 +129,7 @@ def to_raw(self):
             return self._raw
 
         if self._tensor is not None:
-            array = self._tensor.cpu().detach().numpy()
+            array = self._tensor.detach().numpy()
```
Comment (gante) [inline src/transformers/agents/agent_types.py:132]: `numpy()` behaves differently depending on whether it is on CPU or not, so we want to keep the original here ([docs](https://pytorch.org/docs/stable/generated/torch.Tensor.numpy.html#torch.Tensor.numpy))

cases with `.cpu()` and `.numpy()` should be kept as is -- there are many examples of this changes in this PR

File: tests/models/encodec/test_modeling_encodec.py [hunk 3]
```diff
@@ -117,7 +117,7 @@ def prepare_config_and_inputs_for_normalization(self):
         config.normalize = True
 
         processor = EncodecFeatureExtractor(feature_size=config.audio_channels, sampling_rate=config.sampling_rate)
-        input_values = list(input_values.cpu().numpy())
+        input_values = list(input_values.numpy())
```
Comment (gante) [inline tests/models/encodec/test_modeling_encodec.py:120]:
```suggestion
        input_values = input_values.tolist()
```

**Review by gante — COMMENTED** [review]
Thank you for opening the PR!

The changes look good to me, minus the .cpu() + .numpy() change pattern (see comment below)

File: src/transformers/agents/agent_types.py (hunk 2 above)
Comment (cyyever) [reply src/transformers/agents/agent_types.py]: Reverted

**Comment by cyyever** [comment]
@gante Done

**Review by Rocketknight1 — APPROVED** [review]
Yep, this all looks good to me. Thank you!

**Comment by ydshieh** [comment]
The changes in this fail  quite a lot of test, see [this](https://huggingface.co/datasets/hf-internal-testing/transformers_daily_ci/raw/37a90cd406e891e58865aaa4d5d192a0eef01207/2025-04-01/ci_results_run_models_gpu/new_model_failures_with_bad_commit_grouped_by_authors.json). You can search "pr_number": 37060,` to find the failing tests related to this PR.

The job run page is at https://github.com/huggingface/transformers/actions/runs/14186033478

@cyyever @Rocketknight1 I am considering to revert this PR and give @cyyever some time to address the failing tests.

**Comment by ydshieh** [comment]
well, daily CI is on GPU and have a more complete set of tests ...

**Comment by cyyever** [comment]
Can you reopen it?

**Comment by ydshieh** [comment]
You mean reopen the merged PR? It seems impossible for a merged one.

**Comment by ydshieh** [comment]
BTW, let me check something first before you dive iinto it.

**Comment by ydshieh** [comment]
our bad: it's not from this PR. The report has some issue. Sorry to bother you @cyyever

**Comment by cyyever** [comment]
Never mind, we all hope the library becomes better.
